### PR TITLE
Tests for the Session API

### PR DESCRIPTION
**Pull request by weswigham**
Additionally, I added visibility modifiers to all the members, so we declare public as little as is required.

When merged, this should close issue #3813.

### CODE REVIEW

File: tests/cases/unittests/session.ts [hunk 1]
```diff
@@ -0,0 +1,459 @@
+/// <reference path="..\..\..\src\harness\harness.ts" />
+
+module ts.server {
+	let lastWrittenToHost: string,
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Please use 4 spaces instead of tabs.

File: tests/cases/unittests/session.ts [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+			resolvePath: () => void 0,
+			fileExists: () => false,
+			directoryExists: () => false,
+			createDirectory: () => void 0,
+			getExecutingFilePath: () => void 0,
+			getCurrentDirectory: () => void 0,
+			readDirectory: () => void 0,
+			exit: () => void 0
+		},
+		InProcSession = class extends Session {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Is there any reason you don't just use a class declaration here?

File: tests/cases/unittests/session.ts [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+				expect(responses).to.equal(2);
+				expect(resp.body).to.deep.equal(toEcho);
+			});
+			
+			// Queue a configure command
+			cli.execute('configure', {
+				hostInfo: 'unit test',
+				formatOptions: {
+					newLineCharacter: '`n'
+				}
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Is this intended to be ``'`n'``

File: tests/cases/unittests/session.ts [hunk 4]
```diff
@@ -0,0 +1,459 @@
...
+				if (response) {
+					this.output(response, msg.command, msg.seq);
+				}
+			}
+
+			consumeQueue() {
+				while (this.queue.length > 0) {
+					let elem = this.queue[this.queue.length - 1];
+					this.queue = this.queue.slice(0, this.queue.length - 1);
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]:
```TypeScript
let elem = this.queue.pop();
```

File: tests/cases/unittests/session.ts [hunk 5]
```diff
@@ -0,0 +1,459 @@
...
+        perftrc(s: string): void {},
+        info(s: string): void {},
+        startGroup(): void {},
+        endGroup(): void {},
+        msg(s: string, type?: string): void {},
+	};
+	
+	describe('the Session class', () => {
+		let session: Session,
+			lastSent: protocol.Message;
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Use separate variable declarations.

File: tests/cases/unittests/session.ts [hunk 6]
```diff
@@ -0,0 +1,459 @@
...
+		beforeEach(() => {
+			session = new Session(mockHost, Buffer.byteLength, process.hrtime, mockLogger);
+			session.send = (msg: protocol.Message) => {
+				lastSent = msg;
+			};
+		});
+
+		describe('executeCommand', () => {
+			it('should throw when commands are executed with invalid arguments', () => {
+				let req : protocol.FileRequest = {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Our coding style omits spaces preceding the colon.

File: tests/cases/unittests/session.ts [hunk 7]
```diff
@@ -0,0 +1,459 @@
...
+			it('is a noop which can be handled by subclasses', () => {
+				session.exit(); // Does nothing, should keep running tests
+				expect(session).to.exist;
+			});
+		});
+
+		describe('send', () => {
+			it('is an overrideable handle which sends protocol messages over the wire', () => {
+				let msg = {seq: 0, type: 'none'},
+				strmsg = JSON.stringify(msg),
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Use a separate variable statement, and I'd just do `let msg = JSON.stringify({seq: 0, type: "none" });`

File: tests/cases/unittests/session.ts [hunk 8]
```diff
@@ -0,0 +1,459 @@
...
+						command: name,
+						seq: i++,
+						type: 'command'
+					};
+					session.onMessage(JSON.stringify(req));
+					req.seq += 2;
+					req.arguments = {};
+					session.onMessage(JSON.stringify(req));
+					req.seq += 2;
+					req.arguments = null;
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: We use `undefined`, not `null`.

File: tests/cases/unittests/session.ts [hunk 9]
```diff
@@ -0,0 +1,459 @@
+/// <reference path="..\..\..\src\harness\harness.ts" />
+
+module ts.server {
+	let lastWrittenToHost: string,
+	mockHost: ServerHost = {
+		args: [],
+		newLine: '\n',
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Could you use double quotes as per our coding guidelines?

File: tests/cases/unittests/session.ts [hunk 10]
```diff
@@ -0,0 +1,459 @@
+/// <reference path="..\..\..\src\harness\harness.ts" />
+
+module ts.server {
+	let lastWrittenToHost: string,
+	mockHost: ServerHost = {
+		args: [],
+		newLine: '\n',
+		useCaseSensitiveFileNames: true,
+		write: (s) => lastWrittenToHost = s,
+		readFile: () => void 0,
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Just use `undefined`.

File: tests/cases/unittests/session.ts [hunk 11]
```diff
@@ -0,0 +1,459 @@
...
+		writeFile: () => void 0,
+		resolvePath: () => void 0,
+		fileExists: () => false,
+		directoryExists: () => false,
+		createDirectory: () => void 0,
+		getExecutingFilePath: () => void 0,
+		getCurrentDirectory: () => void 0,
+		readDirectory: () => void 0,
+		exit: () => void 0
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Just use a block body for all of these `void 0` functions. 

In fact, you can write them all in method syntax.

```TypeScript
exit() {
}
```

File: tests/cases/unittests/session.ts [hunk 12]
```diff
@@ -0,0 +1,459 @@
+/// <reference path="..\..\..\src\harness\harness.ts" />
+
+module ts.server {
+	let lastWrittenToHost: string,
+	mockHost: ServerHost = {
+		args: [],
+		newLine: '\n',
+		useCaseSensitiveFileNames: true,
+		write: (s) => lastWrittenToHost = s,
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Use a block body here, or else you are always returning `s`.

**Comment by mhegazy** [comment]
looks good to me, along with @DanielRosenwasser's comments. @zhengbli can you take a look as well.

File: tests/cases/unittests/session.ts (hunk 1 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Darn, that was unintentional. I hate tabs. Can we modify the tslint to pick up on this, too?

File: tests/cases/unittests/session.ts (hunk 2 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Class declarations can't be done within a closure and since they're one and done classes, class expressions seem to make more sense than hoisting the declarations out of the test they're for.

Plus, class expressions are a cool new feature we support that we can use when they make sense.

File: tests/cases/unittests/session.ts (hunk 3 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Yep. Trivia: That's the newline character in powershell string literals (go figure). 

The exact value doesn't actually matter for the test (it just checks that the options are sent and responded to), and as far as I know the service isn't supposed to choke on values that aren't `\n` or `\r\n`, so it should be okay.

File: tests/cases/unittests/session.ts (hunk 5 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Is it possible for us to add this one to the tslint for the future, too? It should be able to catch this.

File: tests/cases/unittests/session.ts (hunk 6 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Also missed by a pass of tslint. Would like to see this one captured, too.

File: tests/cases/unittests/session.ts (hunk 8 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: I'm not necessarily checking for expected values here, I'm trying to check that the service doesn't throw when you pass invalid things (or anything) to it 'over the wire'. I should probably actually try passing both (And I think I do - the first omits `arguments` which is equivalent to setting it to `undefined` afaik?)

I should also probably add a test for passing garbage strings to onMessage, rather than just json with invalid/empty arguments.

File: tests/cases/unittests/session.ts (hunk 9 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Again, will do - but tslint didn't capture this style guideline when I ran it over the file. It'd be nice to see it added for the future.

File: tests/cases/unittests/session.ts (hunk 12 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: The TS emit doesn't change to stop that when the function is typed to return void? Huh. I kinda assumed that was a thing. Will do.

File: tests/cases/unittests/session.ts (hunk 11 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Got it. In my own code I've adopted `() => void 0` as an amazingly short noop function that is visually less ambiguous than `() => {}` (which for those less informed about lambdas looks like returning an empty object) and _very_ similar to its type definition: `() => void`. The preference simply bled through here.

File: tests/cases/unittests/session.ts (hunk 7 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: I use both the object message and the stringified message later, I can't simply inline it (without duplicating or undoing the serialization, ofc) - lines 159 and 160 use the string and line 164 uses the object.
I'll split the initialization up as before, and again, this should probably be a tslint rule?

**Comment by weswigham** [comment]
Re: All my responses mentioning tslint - on further investigation most of them _should_ have been picked up by out current tslint settings according to the tslint documentation, but these ones didn't seem to get reported when I ran a pass.

I looked into it a bit and found the following:
`quotemark` and `indent` are both just `true` - the docs says this only requires _consistency_ so when you only run tslint on a single file.... it doesn't really help much in the project-wide sense. 
`indent` should probably have `[true, "spaces"]` added to its argument list ([Found that gem in a PR, the docs don't seem to have been updated with it](https://github.com/palantir/tslint/pull/234)) - but [actually looking over the source, only passing true means it does nothing to check indents at all](https://github.com/palantir/tslint/blob/master/src/rules/indentRule.ts)). Similarly, `"quotemark"` should actually bet set to `[true, "double"]` rather than simply `true` to actually be on, [looking at that rule's source it's not actually run without either "single" or "double" specified.](https://github.com/palantir/tslint/blob/3b3c56cd9c6f114d28f5b16dfec7df161132e0cc/src/rules/quotemarkRule.ts).
`"typedef-whitespace": "nospace"` should be added to the tslint to capture the error of having a space before the colon. 
And I'm not sure tslint has a rule covering multiple declarations at present... I know jslint can (and, by default, has you combine them, hence my style). May need to request/make a new rule to capture that?

File: tests/cases/unittests/session.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: > Class declarations can't be done within a closure

This is untrue, and if it is, that's a bug. You can have class declarations in functions, so let's do that here.

File: tests/cases/unittests/session.ts [hunk 13]
```diff
@@ -0,0 +1,459 @@
...
+			consumeQueue() {
+				while (this.queue.length > 0) {
+					let elem = this.queue[this.queue.length - 1];
+					this.queue = this.queue.slice(0, this.queue.length - 1);
+					this.handleRequest(elem);
+				}
+			}
+		},
+		InProcClient = class {
+			private server: Session & {enqueue: (msg: protocol.Request) => void};
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Could you instead use a type alias here?

File: tests/cases/unittests/session.ts [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+					responseRequired: true
+				}));
+			}
+			
+			send(msg: protocol.Message) {
+				this.client.handle(msg);
+			}
+
+			enqueue(msg: protocol.Request) {
+				this.queue = [msg].concat(this.queue);
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]:
```TypeScript
this.queue.unshift(msg);
```

File: tests/cases/unittests/session.ts [hunk 15]
```diff
@@ -0,0 +1,459 @@
...
+			}
+			
+			handleRequest(msg: protocol.Request) {
+				let response: protocol.Response;
+				try {
+					response = this.executeCommand(msg).response;
+				} catch (e) {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: `catch` on the next line

File: tests/cases/unittests/session.ts [hunk 16]
```diff
@@ -0,0 +1,459 @@
...
+					this.output(undefined, msg.command, msg.seq, e.toString());
+					return;
+				}
+				if (response) {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Does `executeCommand` potentially not have a response?

File: tests/cases/unittests/session.ts [hunk 17]
```diff
@@ -0,0 +1,459 @@
...
+			private eventHandlers: ts.Map<(args: any) => void> = {};
+
+			handle(msg: protocol.Message): void {
+				if (msg.type === 'response') {
+					let response = <protocol.Response>msg;
+					if (this.callbacks[response.request_seq]) {
+						this.callbacks[response.request_seq](response);
+						delete this.callbacks[response.request_seq];
+					}
+				} else if (msg.type === 'event') {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: `else` on the next line

File: tests/cases/unittests/session.ts [hunk 18]
```diff
@@ -0,0 +1,459 @@
...
+					let event = <protocol.Event>msg;
+					this.emit(event.event, event.body);
+				}
+			}
+
+			emit(name: string, args: any): void {
+				if (!this.eventHandlers[name]) {
+					return;
+				}
+				this.eventHandlers[name](args);
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Remove the `!` from the condition and use this in place of `return`

File: tests/cases/unittests/session.ts (hunk 14 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: :+1:

File: tests/cases/unittests/session.ts [hunk 19]
```diff
@@ -0,0 +1,459 @@
...
+			directoryExists: () => false,
+			createDirectory: () => void 0,
+			getExecutingFilePath: () => void 0,
+			getCurrentDirectory: () => void 0,
+			readDirectory: () => void 0,
+			exit: () => void 0
+		},
+		InProcSession = class extends Session {
+			private queue: protocol.Request[] = [];
+			constructor(private client: {handle: (msg: protocol.Message) => void}) {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: When we convert the class expressions to class declarations, use the class name in place of the type literal.

File: tests/cases/unittests/session.ts (hunk 13 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: TS doesn't allow local type aliases and I'd been trying to keep all the types internal to the test. So I'd have to hoist the alias all the way out to the ambient level despite no other tests or types in the module depending on it. Eh?

Makes me want local type aliases.

File: tests/cases/unittests/session.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Yup, that's a good idea.

File: tests/cases/unittests/session.ts (hunk 7 above)
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Sorry, missed that, no worries.

File: tests/cases/unittests/session.ts (hunk 15 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Will do. And there's already [an open issue](https://github.com/palantir/tslint/issues/88) in tslint to add support for a rule to check this.

File: tests/cases/unittests/session.ts (hunk 16 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Yes. `response` can be `undefined` while `responseRequired` is true. (Indicating that it should pretty much just ack the message.) This is done in the server.

File: tests/cases/unittests/session.ts (hunk 17 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Will do. Same as for the `catch` above.

File: tests/cases/unittests/session.ts (hunk 18 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Will do. It's my habit to use fail fast if statements to reduce nesting - in this case there's no conditions to nest so it didn't really matter.

File: tests/cases/unittests/session.ts (hunk 19 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: To be fair, if they were kept as class expressions I realize now that I could probably just use `typeof ClassExpressionVarName`. Didn't occur to me at the time.

File: tests/cases/unittests/session.ts (hunk 2 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: `class declarations are only supported directly inside a module or as a top-level declaration` at least that's the error on a TS playground (and likely in VS Code (which is what I use most of the time) as well)  - If this isn't true for ToT TS then it's my fault for trusting my editor too much. (I did explicitly ignore its errors about class expressions not being allowed since I had read over the PR adding them.) 

I have yet to set up Code in such a way where is can track the latest TS while I'm editing the latest TS, so it may be my fault for thinking that. Actual class declarations are way more idiomatic here if they're actually allowed in that context.

File: tests/cases/unittests/session.ts (hunk 4 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: :+1:

File: tests/cases/unittests/session.ts (hunk 2 above)
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: That's odd though, you should also get something like

```
'class' expressions are not currently supported.
```

Let's talk tomorrow regarding swapping out language services in VSCode. I haven't done it, but we can look into it.

**Comment by DanielRosenwasser** [comment]
@weswigham thanks for pointing out the lack of TSLint rules, I've created #3994 to track the issue.

File: tests/cases/unittests/session.ts (hunk 2 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: I did get that error and summarily ignored it since I knew they were a thing (and the actual compiler confirmed I was right). I'll actually check against the compiler tomorrow if classes are supported in that context. (And likely they are and I'll make the change.)

Comment (weswigham) [inline tests/cases/unittests/session.ts]: Followup: Just learned how to use VS Code with arbitrary versions of TS. There's a `"typescript.tsdk"` config parameter that I can just point at the built/local folder. Nice. Would it be worth adding this as a workspace setting to the .settings folder in the TS repo?

File: tests/cases/unittests/session.ts [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+                    seq: 0,
+                    body: undefined
+                });
+            });
+        });
+
+        describe("exit", () => {
+            it("is a noop which can be handled by subclasses", () => {
+                session.exit(); // Does nothing, should keep running tests
+                expect(session).to.exist;
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Should this be a method call?

File: tests/cases/unittests/session.ts [hunk 21]
```diff
@@ -0,0 +1,454 @@
+/// <reference path="..\..\..\src\harness\harness.ts" />
+
+module ts.server {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Use the `namespace` keyword here.

File: tests/cases/unittests/session.ts [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+                    success: true
+                });
+            });
+        });
+    });
+
+    describe("how Session is extendable via subclassing", () => {
+        class TestSession extends Session {
+            lastSent: protocol.Message;
+            customHandler: string = "testhandler";
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: You can omit the type annotation

File: tests/cases/unittests/session.ts [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+                while (this.queue.length > 0) {
+                    const elem = this.queue.pop();
+                    this.handleRequest(elem);
+                }
+            }
+        }
+        
+        class InProcClient {
+            private server: InProcSession;
+            private seq: number = 0;
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: You can omit the type annotation

File: tests/cases/unittests/session.ts [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+                }
+            };
+            new ServiceSession();
+        });
+    });
+
+    describe("an example of using the Session API to create an in-process server", () => {
+        class InProcSession extends Session {
+            private queue: protocol.Request[] = [];
+            constructor(private client: {handle: (msg: protocol.Message) => void}) {
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Can we still make this `InProcClient`?

File: tests/cases/unittests/session.ts [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+                });
+            });
+        });
+    });
+
+    describe("how Session is extendable via subclassing", () => {
+        class TestSession extends Session {
+            lastSent: protocol.Message;
+            customHandler: string = "testhandler";
+            constructor(){
```
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: Space after the `)`

**Comment by DanielRosenwasser** [comment]
Just a few more comments to fix up, and I think we should be good to go.

File: tests/cases/unittests/session.ts (hunk 20 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Honestly the line probably shouldn't be there at all... Really, I effectively wanted an `assert(true)` after calling session.exit() to indicate that the process should still be running. I'm not really sure how to test that a function which should be a noop is, in fact, still a noop. (Beyond literally stringifying the function and checking it's source, which seems extremely tautological)

Actually, in every case `Session` gets subclassed - maybe we should just mark Session as `abstract` and `exit()` as an abstract method (then remove the test)? I think that may be more in line with what is supposed to occur.

File: tests/cases/unittests/session.ts (hunk 24 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Oh, yeah. Sorry, I missed that one since it wasn't one of the intersection types.

File: tests/cases/unittests/session.ts (hunk 20 above)
Comment (DanielRosenwasser) [inline tests/cases/unittests/session.ts]: I'd be fine with that as a separate PR. In the mean time, consider removing the test.

File: tests/cases/unittests/session.ts (hunk 22 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Can we craft a tslint rule which checks for unnecessary types on inferrable literals? Seems like something we can do.

File: tests/cases/unittests/session.ts (hunk 21 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Noting that we should add a tslint rule to prefer the new `namespace` keyword over `module`.

File: tests/cases/unittests/session.ts (hunk 20 above)
Comment (weswigham) [inline tests/cases/unittests/session.ts]: Got it.

**Comment by DanielRosenwasser** [comment]
Destler would be proud. Thanks!
